README.md
This package is used to get time duration from now and given time for kurdish, arabic and english, simple and easy to use.

## Features

🌟 Changes date to time ago.

<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/all.png"><br><br>

## Getting started

Depend on it

add package to your pubspec.yaml file
```yaml
    dependecies:
        flutter:
            sdk: flutter
        z_time_ago: any
```

Or

Run this command in terminal
```dart
flutter pub add z_time_ago
```
<br>

Import it

```dart
import 'package:z_time_ago/z_time_ago.dart';
```
<br><br><br>

## Usage

With DateTime Type

```dart
String result = ZTimeAgo().getTimeAgo(
    date: DateTime.parse('2021-01-15 20:50:36.977'),
    language: Language.kurdish,
);
```
<br><br>

With String Type

```dart
String result = ZTimeAgo().getTimeAgo(
    date: '2021-01-15 20:50:36.977',
    language: Language.kurdish,
);
```
<br><br><br><br>

## More Examples
```dart
ZTimeAgo().getTimeAgo(
    date: '2022-01-28 11:46:54.897839',
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/kurdishMonthAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: '2022-01-28 11:46:54.897839',
    language: Language.arabic,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/arabicMonthAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: '2022-03-01 08:46:54.897839',
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/englishMonthAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: '2022-03-01 08:46:54.897839',
    language: Language.kurdish,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/kurdishHoursAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: '2022-03-01 08:46:54.897839',
    language: Language.arabic,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/arabicHoursAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: '2022-03-01 08:46:54.897839',
    language: Language.english,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/englishHoursAgo.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: DateTime.now(),
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/kurdishNow.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: DateTime.now(),
    language: Language.arabic,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/arabicNow.png"><br><br>

```dart
ZTimeAgo().getTimeAgo(
    date: DateTime.now(),
    language: Language.english,
),
```
<img src="https://github.com/zakarya0/z_time_ago/raw/main/assets/englishNow.png"><br><br>

## Bugs And Issues

In case of having any issues, feel free to open an <a href="https://github.com/zakarya0/z_time_ago/issues/new">issue</a>, & if you think that the library needs a feature let me know by adding it in this <a href="https://github.com/zakarya0/z_time_ago/issues/new?template=feature.md">link</a>

CHANGELOG.md
## 0.0.1

package to get time duration from now and given time

## 0.0.3

date parameter accepts timestamps(String) not just DateTime

## 0.1.2

fixing some bugs

lib/z_time_ago.dart
library z_time_ago;

import 'package:flutter/material.dart';
import 'src/data.dart';

enum Language {
  arabic,
  english,
  kurdish,
}

class ZTimeAgo {
  String getTimeAgo({
    required dynamic date,
    Language language = Language.kurdish,
  }) {
    DateTime inputDate = DateTime.now();
    if (date.runtimeType == String) {
      inputDate = DateTime.parse(date);
    } else if (date.runtimeType == DateTime) {
      inputDate = date;
    } else {
      debugPrint('Enter a valid type');
      return '';
    }
    String message = '';
    var currentDate = DateTime.now();
    var difference = currentDate.difference(inputDate);
    Map data = Data().getData(language);
    if (difference.inSeconds < 0) {
      debugPrint('the date you entered is after current date');
      return '';
    } else {
      if (difference.inSeconds < 60) {
        message = data['keys']['now'];
      } else if (difference.inSeconds >= 60 && difference.inSeconds < 120) {
        message = data['keys']['aMinuteAgo'];
      } else if (difference.inMinutes >= 1 && difference.inMinutes < 60) {
        if (language == Language.arabic) {
          message =
              '${data['keys']['prefix']} ${data['numbers']['${difference.inMinutes}']} ${data['keys']['minutesAgo']}';
        } else {
          message =
              '${data['numbers']['${difference.inMinutes}']} ${data['keys']['minutesAgo']}';
        }
      } else if (difference.inHours == 1) {
        message = data['keys']['anHourAgo'];
      } else if (difference.inHours >= 1 && difference.inHours < 24) {
        if (language == Language.arabic) {
          message =
              '${data['keys']['prefix']} ${data['numbers']['${difference.inHours}']} ${data['keys']['hoursAgo']}';
        } else {
          message =
              '${data['numbers']['${difference.inHours}']} ${data['keys']['hoursAgo']}';
        }
      } else if (difference.inDays == 1) {
        message = data['keys']['aDayAgo'];
      } else if (difference.inDays >= 2 && difference.inDays < 7) {
        if (language == Language.arabic) {
          message =
              '${data['keys']['prefix']} ${data['numbers']['${difference.inDays}']} ${data['keys']['daysAgo']}';
        } else {
          message =
              '${data['numbers']['${difference.inDays}']} ${data['keys']['daysAgo']}';
        }
      } else if (difference.inDays >= 7 && difference.inDays < 14) {
        message = data['keys']['aWeekAgo'];
      } else if (difference.inDays >= 14 && difference.inDays < 30) {
        if (language == Language.arabic) {
          message =
              '${data['keys']['prefix']} ${data['numbers']['${difference.inDays ~/ 7.toInt()}']} ${data['keys']['weeksAgo']}';
        } else {
          message =
              '${data['numbers']['${difference.inDays ~/ 7.toInt()}']} ${data['keys']['weeksAgo']}';
        }
      } else if (difference.inDays >= 30 && difference.inDays < 60) {
        message = data['keys']['aMonthAgo'];
      } else if (difference.inDays >= 60 && difference.inDays < 365) {
        if (language == Language.arabic) {
          message =
              '${data['keys']['prefix']} ${data['numbers']['${difference.inDays ~/ 30.toInt()}']} ${data['keys']['monthsAgo']}';
        } else {
          message =
              '${data['numbers']['${difference.inDays ~/ 30.toInt()}']} ${data['keys']['monthsAgo']}';
        }
      } else if (difference.inDays >= 365 && difference.inDays < 731) {
        message = data['keys']['aYearAgo'];
      } else {
        String finalDate = inputDate.toString().split(' ')[0];
        var splittedDate = finalDate.split('-');
        String day = '';
        String month = '';
        String year = '';
        if (splittedDate[2][0] == '0') {
          day = data['numbers']['0'] + data['numbers'][splittedDate[2][1]];
        } else {
          day = data['numbers'][splittedDate[2]];
        }
        if (splittedDate[1][0] == '0') {
          month = data['numbers']['0'] + data['numbers'][splittedDate[1][1]];
        } else {
          month = data['numbers'][splittedDate[1]];
        }
        String firstPartOfYear =
            data['numbers'][splittedDate[0].substring(0, 2)];
        String secondPartOfYear = splittedDate[0].substring(2, 4);
        if (secondPartOfYear[0] == '0') {
          secondPartOfYear =
              data['numbers']['0'] + data['numbers'][secondPartOfYear[1]];
        } else {
          secondPartOfYear = data['numbers'][splittedDate[0].substring(2, 4)];
        }
        year = '$firstPartOfYear$secondPartOfYear';
        message = '$year-$month-$day';
      }
      return message;
    }
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:z_time_ago/z_time_ago.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Time Ago',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Time Ago'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String timeAgo = '';

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'click the button to get time ago',
            ),
            Text(
              timeAgo,
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            timeAgo = ZTimeAgo().getTimeAgo(
              date: DateTime.parse('2021-02-15 20:50:36.977'),
              language: Language.english,
            );
          });
        },
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/src/data.dart
import '../z_time_ago.dart';

class Data {
  final _english = {
    "keys": {
      "now": "just now",
      "aMinuteAgo": "a minute ago",
      "minutesAgo": "minutes ago",
      "anHourAgo": "an hour ago",
      "hoursAgo": "hours ago",
      "aDayAgo": "a day ago",
      "daysAgo": "days ago",
      "aWeekAgo": "a week ago",
      "weeksAgo": "weeks ago",
      "aMonthAgo": "a month ago",
      "monthsAgo": "months ago",
      "aYearAgo": "a year ago"
    },
    "numbers": {
      "0": "0",
      "1": "1",
      "2": "2",
      "3": "3",
      "4": "4",
      "5": "5",
      "6": "6",
      "7": "7",
      "8": "8",
      "9": "9",
      "10": "10",
      "11": "11",
      "12": "12",
      "13": "13",
      "14": "14",
      "15": "15",
      "16": "16",
      "17": "17",
      "18": "18",
      "19": "19",
      "20": "20",
      "21": "21",
      "22": "22",
      "23": "23",
      "24": "24",
      "25": "25",
      "26": "26",
      "27": "27",
      "28": "28",
      "29": "29",
      "30": "30",
      "31": "31",
      "32": "32",
      "33": "33",
      "34": "34",
      "35": "35",
      "36": "36",
      "37": "37",
      "38": "38",
      "39": "39",
      "40": "40",
      "41": "41",
      "42": "42",
      "43": "43",
      "44": "44",
      "45": "45",
      "46": "46",
      "47": "47",
      "48": "48",
      "49": "49",
      "50": "50",
      "51": "51",
      "52": "52",
      "53": "53",
      "54": "54",
      "55": "55",
      "56": "56",
      "57": "57",
      "58": "58",
      "59": "59",
      "60": "60"
    }
  };

  final _arabic = {
    "keys": {
      "prefix": "منذ",
      "now": "الآن",
      "aMinuteAgo": "منذ دقيقة",
      "minutesAgo": "دقائق",
      "anHourAgo": "منذ ساعة",
      "hoursAgo": "ساعات",
      "aDayAgo": "منذ یوم",
      "daysAgo": "ایام",
      "aWeekAgo": "منذ أسبوع",
      "weeksAgo": "أسابيع",
      "aMonthAgo": "منذ شهر",
      "monthsAgo": "اشهر",
      "aYearAgo": "منذ عام"
    },
    "numbers": {
      "0": "٠",
      "1": "١",
      "2": "٢",
      "3": "٣",
      "4": "٤",
      "5": "٥",
      "6": "٦",
      "7": "٧",
      "8": "٨",
      "9": "٩",
      "10": "١٠",
      "11": "١١",
      "12": "١٢",
      "13": "١٣",
      "14": "١٤",
      "15": "١٥",
      "16": "١٦",
      "17": "١٧",
      "18": "١٨",
      "19": "١٩",
      "20": "٢٠",
      "21": "٢١",
      "22": "٢٢",
      "23": "٢٣",
      "24": "٢٤",
      "25": "٢٥",
      "26": "٢٦",
      "27": "٢٧",
      "28": "٢٨",
      "29": "٢٩",
      "30": "٣٠",
      "31": "٣١",
      "32": "٣٢",
      "33": "٣٣",
      "34": "٣٤",
      "35": "٣٥",
      "36": "٣٦",
      "37": "٣٧",
      "38": "٣٨",
      "39": "٣٩",
      "40": "٤٠",
      "41": "٤١",
      "42": "٤٢",
      "43": "٤٣",
      "44": "٤٤",
      "45": "٤٥",
      "46": "٤٦",
      "47": "٤٧",
      "48": "٤٨",
      "49": "٤٩",
      "50": "٥٠",
      "51": "٥١",
      "52": "٥٢",
      "53": "٥٣",
      "54": "٥٤",
      "55": "٥٥",
      "56": "٥٦",
      "57": "٥٧",
      "58": "٥٨",
      "59": "٥٩",
      "60": "٦٠"
    }
  };

  final _kurdish = {
    "keys": {
      "now": "ئێستا",
      "aMinuteAgo": "١ خولەک لەمەوبەر",
      "minutesAgo": "خولەک لەمەوبەر",
      "anHourAgo": "١ کاتژمێر لەمەوبەر",
      "hoursAgo": "کاتژمێر لەمەوبەر",
      "aDayAgo": "١ ڕۆژ لەمەوبەر",
      "daysAgo": "ڕۆژ لەمەوبەر",
      "aWeekAgo": "١ هەفتە لەمەوبەر",
      "weeksAgo": "هەفتە لەمەوبەر",
      "aMonthAgo": "١ مانگ لەمەوبەر",
      "monthsAgo": "مانگ لەمەوبەر",
      "aYearAgo": "١ ساڵ لەمەوبەر"
    },
    "numbers": {
      "0": "٠",
      "1": "١",
      "2": "٢",
      "3": "٣",
      "4": "٤",
      "5": "٥",
      "6": "٦",
      "7": "٧",
      "8": "٨",
      "9": "٩",
      "10": "١٠",
      "11": "١١",
      "12": "١٢",
      "13": "١٣",
      "14": "١٤",
      "15": "١٥",
      "16": "١٦",
      "17": "١٧",
      "18": "١٨",
      "19": "١٩",
      "20": "٢٠",
      "21": "٢١",
      "22": "٢٢",
      "23": "٢٣",
      "24": "٢٤",
      "25": "٢٥",
      "26": "٢٦",
      "27": "٢٧",
      "28": "٢٨",
      "29": "٢٩",
      "30": "٣٠",
      "31": "٣١",
      "32": "٣٢",
      "33": "٣٣",
      "34": "٣٤",
      "35": "٣٥",
      "36": "٣٦",
      "37": "٣٧",
      "38": "٣٨",
      "39": "٣٩",
      "40": "٤٠",
      "41": "٤١",
      "42": "٤٢",
      "43": "٤٣",
      "44": "٤٤",
      "45": "٤٥",
      "46": "٤٦",
      "47": "٤٧",
      "48": "٤٨",
      "49": "٤٩",
      "50": "٥٠",
      "51": "٥١",
      "52": "٥٢",
      "53": "٥٣",
      "54": "٥٤",
      "55": "٥٥",
      "56": "٥٦",
      "57": "٥٧",
      "58": "٥٨",
      "59": "٥٩",
      "60": "٦٠"
    }
  };

  getData(Language language) {
    if (language == Language.english) {
      return _english;
    } else if (language == Language.arabic) {
      return _arabic;
    } else {
      return _kurdish;
    }
  }
}
